### PR TITLE
Provide new trace query protocol

### PR DESCRIPTION
**Pull request by wu-sheng**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided

### New feature or improvement
- Describe the details and related test reports.
Based on I and @hanahmily 's discussion, we decided to refactor the whole trace query page. For supporting that, backend should open the Segment data structure to frontend, in order to give front end more freedom to focus on visualization.

So, in this pull request, I will submit the new return value of trace query service. Please review the protocol from here:
```
type Trace {
  segments: [Segment!]!
  # Segment IDs in refs may not exist in #segments list, because MQ batch process trace segments didn't share the same ID.
  # UI should add visual node for non-exist segment.
  refs: [SegmentRef!]!
}
```

### CODE REVIEW

**Comment by wu-sheng** [comment]
@peng-yongsheng @hanahmily I finished the new codes. But no test yet.

**Comment by wu-sheng** [comment]
@apache/skywalking-committers Please be advised, this is break feature, after this is merged, query graphQL is not compatible with 5.0.0-alpha UI any more.

File: apm-protocol/apm-ui-protocol/src/main/resources/ui-graphql/trace.graphqls [hunk 1]
```diff
@@ -52,16 +52,48 @@ enum QueryOrder {
 
 # The trace represents a distributed trace, includes all segments and spans.
 type Trace {
+  segments: [Segment!]!
+  # Segment IDs in refs may not exist in #segments list, because MQ batch process trace segments didn't share the same ID.
+  # UI should add visual node for non-exist segment.
+  refs: [SegmentRef!]!
+}
+
+type Segment {
```
Comment (peng-yongsheng) [inline apm-protocol/apm-ui-protocol/src/main/resources/ui-graphql/trace.graphqls]: Some attribute need explain, e.g. name, brief.

**Comment by wu-sheng** [comment]
@peng-yongsheng Comments have been added now.

**Comment by wu-sheng** [comment]
@peng-yongsheng  I shut down the short column name mechanism to run the tests. 
- Request
```


{
  queryTrace(traceId: "1.2.3") {
    segments{
      segmentId
      applicationCode
      name
      brief{
        http{
          count
          duration
        }
        rpc{
          count
          duration
        }
        mq{
          count
          duration
        }
        db{
          count
          duration
        }
        cache{
          count
          duration
        }
      }
      spans{
        spanId
        parentSpanId
        refs{
          traceId
          parentSegmentId
          parentSpanId
          type
        }
        startTime
        endTime
        operationName
        type
        peer
        component
        isError
        layer
        tags{
          key
          value
        }
        logs{
          time
          data{
            key 
            value
          }
        }
      }
      isError
      duration
    }
    refs{
      sourceSegment
      targetSegment
      callType
    }
  }
}
```

- Result
```
{
  "data": {
    "queryTrace": {
      "segments": [
        {
          "segmentId": "7.8.9",
          "applicationCode": "dubbox-provider",
          "name": "org.skywaking.apm.testcase.dubbo.services.GreetService.doBusiness()",
          "brief": {
            "http": {
              "count": 0,
              "duration": 0
            },
            "rpc": {
              "count": 0,
              "duration": 1000
            },
            "mq": {
              "count": 0,
              "duration": 0
            },
            "db": {
              "count": 0,
              "duration": 980
            },
            "cache": {
              "count": 0,
              "duration": 0
            }
          },
          "spans": [
            {
              "spanId": 1,
              "parentSpanId": 0,
              "refs": [],
              "startTime": 1524037537349,
              "endTime": 1524037538329,
              "operationName": "mongodb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]",
              "type": "Exit",
              "peer": "localhost:27017",
              "component": "MongoDB",
              "isError": false,
              "layer": "Database",
              "tags": [],
              "logs": []
            },
            {
              "spanId": 0,
              "parentSpanId": -1,
              "refs": [
                {
                  "traceId": "1.2.3",
                  "parentSegmentId": "4.5.6",
                  "parentSpanId": 1,
                  "type": "CROSS_PROCESS"
                }
              ],
              "startTime": 1524037537339,
              "endTime": 1524037538339,
              "operationName": "org.skywaking.apm.testcase.dubbo.services.GreetService.doBusiness()",
              "type": "Entry",
              "peer": "",
              "component": "Dubbo",
              "isError": false,
              "layer": "RPCFramework",
              "tags": [],
              "logs": []
            }
          ],
          "isError": false,
          "duration": 1524037536359
        },
        {
          "segmentId": "4.5.6",
          "applicationCode": "dubbox-consumer",
          "name": "/dubbox-case/case/dubbox-rest",
          "brief": {
            "http": {
              "count": 0,
              "duration": 2000
            },
            "rpc": {
              "count": 0,
              "duration": 1980
            },
            "mq": {
              "count": 0,
              "duration": 0
            },
            "db": {
              "count": 0,
              "duration": 0
            },
            "cache": {
              "count": 0,
              "duration": 0
            }
          },
          "spans": [
            {
              "spanId": 1,
              "parentSpanId": 0,
              "refs": [],
              "startTime": 1524037536849,
              "endTime": 1524037538829,
              "operationName": "org.skywaking.apm.testcase.dubbo.services.GreetService.doBusiness()",
              "type": "Exit",
              "peer": "172.25.0.4:20880",
              "component": "Dubbo",
              "isError": false,
              "layer": "RPCFramework",
              "tags": [],
              "logs": []
            },
            {
              "spanId": 0,
              "parentSpanId": -1,
              "refs": [],
              "startTime": 1524037536839,
              "endTime": 1524037538839,
              "operationName": "/dubbox-case/case/dubbox-rest",
              "type": "Entry",
              "peer": "",
              "component": "Tomcat",
              "isError": false,
              "layer": "Http",
              "tags": [],
              "logs": []
            }
          ],
          "isError": false,
          "duration": 1524037534859
        }
      ],
      "refs": [
        {
          "sourceSegment": "7.8.9",
          "targetSegment": "4.5.6",
          "callType": "RPCFramework"
        }
      ]
    }
  }
}
```

FYI @hanahmily and @peng-yongsheng This test is only based on MockAgent, so there is no tags and logs in those segments.

**Comment by wu-sheng** [comment]
@peng-yongsheng I will put this feature into **RC** because @hanahmily asked. The new UI design is facing some blockings, some UI components need rewrite.

**Comment by wu-sheng** [comment]
Cancel this.

**Pull request closed without merging**